portal/src/graphql/portal/DNSConfigurationScreen.tsx

```diff
@@ -0,0 +1,598 @@
+import { PortalAPIAppConfig } from "../../types";
+import ShowError from "../../ShowError";
+import ShowLoading from "../../ShowLoading";
+import ButtonWithLoading from "../../ButtonWithLoading";
+import { actionButtonTheme, destructiveTheme } from "../../theme";
+import { useDropdown, useTextField } from "../../hook/useInput";
+import {
+  GenericErrorHandlingRule,
+  useGenericError,
+} from "../../error/useGenericError";
+import ErrorDialog from "../../error/ErrorDialog";
+import { clearEmptyObject } from "../../util/misc";
+
+import styles from "./DNSConfigurationScreen.module.scss";
+
+interface DNSConfigurationProps {
+  domains: Domain[];
+  rawAppConfig: PortalAPIAppConfig | null;
+  effectiveAppConfig: PortalAPIAppConfig | null;
+}
+
+interface PublicOriginConfigurationProps {
+  rawAppConfig: PortalAPIAppConfig | null;
+  effectiveAppConfig: PortalAPIAppConfig | null;
+  verifiedDomains: Domain[];
+}
+
+interface DomainListItem {
+  id: string;
+  domain: string;
+  isVerified: boolean;
+  isCustom: boolean;
+}
+
+interface DomainListActionButtonsProps {
+  domainID: string;
+  domain: string;
+  isCustomDomain: boolean;
+  isVerified: boolean;
+  onDeleteClick: (domainID: string, domain: string) => void;
+}
+
+interface DeleteDomainDialogData {
+  domainID: string;
+  domain: string;
+}
+
+interface DeleteDomainDialogProps extends Partial<DeleteDomainDialogData> {
+  visible: boolean;
+  dismissDialog: () => void;
+}
+
+const DOMAIN_LIST_STYLES: IDetailsListProps["styles"] = {
+  headerWrapper: { marginTop: "-10px" },
+};
+
+function makeDomainListColumn(renderToString: (messageID: string) => string) {
+  return [
+    {
+      key: "domain",
+      name: renderToString("DNSConfigurationScreen.domain-list.header.domain"),
+      minWidth: 250,
+      className: styles.domainListColumn,
+    },
+    {
+      key: "isVerified",
+      name: renderToString("DNSConfigurationScreen.domain-list.header.status"),
+      minWidth: 100,
+      className: styles.domainListColumn,
+    },
+    {
+      key: "action",
+      name: renderToString("action"),
+      minWidth: 150,
+      className: styles.domainListColumn,
+    },
+  ];
+}
+
+const PublicOriginConfiguration: React.FC<PublicOriginConfigurationProps> = function PublicOriginConfiguration(
+  props: PublicOriginConfigurationProps
+) {
+  const { rawAppConfig, effectiveAppConfig, verifiedDomains } = props;
+  const { appID } = useParams();
+
+  const {
+    updateAppConfig,
+    loading: updatingAppConfig,
+    error: updateAppConfigError,
+  } = useUpdateAppConfigMutation(appID);
+
+  const initialPublicOrigin = useMemo(() => {
+    return effectiveAppConfig?.http?.public_origin ?? "";
+  }, [effectiveAppConfig]);
+
+  const publicOriginOptionKeys = useMemo(() => {
+    const keys = verifiedDomains.map((domain) => {
+      try {
+        // assume scheme included if success
+        const url = new URL(domain.domain);
```

Comment:
This should always throw an error, since a domain without scheme is not a valid URL. Instead, just appending the `https://` prefix should be good enough for our purpose.

Reply:
Then we need to add verification to make sure user cannot add domain with scheme @kiootic

Reply:
After the changes in #504, adding an invalid domain should properly produce an API error with reason `InvalidDomain`.